### DISCUSSION TITLE
Cleaner/more idiomatic way to shape queries with joins

### DISCUSSION THREAD

**Discussion by caromo** [slick/slick#2437] [Questions] [answered]
Hello,

scastie for reference: https://scastie.scala-lang.org/IivI7nLtSOatbIWZhFUn8w

I am trying to build a one-to-many relationship in Slick, and while the querying part is easy enough I was having trouble with shaping the result to be the outside class that the rest of my code uses. I tried using a groupBy and modifying the inner tuples but I am not sure if this is the 'standard' way to handle cases like these.

Three main questions arise from my attempt:
1. Is there a more standard/cleaner way to do this transformation
2. Is there a way to transform `TeamModel` and the quota tuples into a `Team` (via an implicit conversion?) 
3. Since I use a wrapper trait `Repository` to handle common functions, is there a way to go about having the common functions return the outer class `Team`? (Maybe related to #2)

**Comment by nafg** [comment]
I'm not so clear what you're asking. Perhaps I could help over a call?

**Reply by nafg** [reply]
Or just clarify the question better?

**Reply by caromo** [reply]
Since I made more than one case class to represent a table + its FKs (`TeamModel`, `TeamQuota`), is there a better way to put the `List[TeamQuota]` with its `TeamModel` other than the way I did in my scastie snippet? (see `tupledJoin` and `convertListToTeam` on line 92) My goal is to end up with a `Team` by merging the Model with the Quotas

Since the query returns a Seq of (`TeamModel`, `TeamQuota`), the way I currently do it does the groupBy outside of the db but I don't know if it is possible to do it inside the db

**Comment by nafg** [comment, marked as answer]
> 2. Is there a way to transform `TeamModel` and the quota tuples into a `Team` (via an implicit conversion?)

I don't think this is a good use case for implicits

> 3. Since I use a wrapper trait `Repository` to handle common functions, is there a way to go about having the common functions return the outer class `Team`? (Maybe related to [Table definitions in companion object result in generation of invalid SQL #2](https://github.com/slick/slick/issues/2))

That issue was closed 12 years ago so I'm not sure its relevance. You can abstract however you want. I'm not clear what the obstacle is.

> Since I made more than one case class to represent a table + its FKs (`TeamModel`, `TeamQuota`), is there a better way to put the `List[TeamQuota]` with its `TeamModel` other than the way I did in my scastie snippet? (see `tupledJoin` and `convertListToTeam` on line 92) My goal is to end up with a `Team` by merging the Model with the Quotas

It looks fine, except that
1. I would not have `Future` be part of the signature of `convertListToTeam`; instead the caller should run it inside `Future#map`.
2. In Scala 2.13+ you can simplify the code with `groupMap`

> Since the query returns a Seq of (`TeamModel`, `TeamQuota`), the way I currently do it does the groupBy outside of the db but I don't know if it is possible to do it inside the db

In general, RDBMSes always return 2D tables, not structured data. So your solution is normal in my experience. That said, there are other options depending on the database. For instance, H2 and Postgres have `ARRAY_AGG`. For Postgres array support you may need to use https://github.com/tminglei/slick-pg. Then you can get the database to return one row per team. But I'm not sure how you'd put the quota pairs into an array. With Postgres you might use jsonb or hstore. Or perhaps you could do one array column for the keys and one for the values. I'm not sure if that would work right or how to make sure. So it might not be worth it.

**Reply by caromo** [reply]
> That issue was closed 12 years ago so I'm not sure its relevance. You can abstract however you want. I'm not clear what the obstacle is.

It was in reference to # 2 on my list of questions, did not realize that github linked to the issues, sorry for the confusion

**Reply by nafg** [reply]
Ha that's funny! (It didn't just hyperlink, it spelled out the issue title)

Anyway is anything still unanswered sufficiently?
